go-controller/pkg/crd/userdefinednetwork/v1/types.go

```diff
@@ -0,0 +1,248 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=userdefinednetworks,scope=Namespaced
+// +kubebuilder:singular=userdefinednetwork
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// UserDefinedNetwork describe network request for a Namespace.
+type UserDefinedNetwork struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Spec is immutable"
+	// +kubebuilder:validation:XValidation:rule="has(self.topology) && self.topology == 'Layer3' ? has(self.layer3): !has(self.layer3)", message="spec.layer3 is required when topology is Layer3 and forbidden otherwise"
+	// +kubebuilder:validation:XValidation:rule="has(self.topology) && self.topology == 'Layer2' ? has(self.layer2): !has(self.layer2)", message="spec.layer2 is required when topology is Layer2 and forbidden otherwise"
+	// +kubebuilder:validation:XValidation:rule="has(self.topology) && self.topology == 'LocalNet' ? has(self.localNet): !has(self.localNet)", message="spec.localNet is required when topology is LocalNet and forbidden otherwise"
+	// +required
+	Spec UserDefinedNetworkSpec `json:"spec"`
+	// +optional
+	Status UserDefinedNetworkStatus `json:"status,omitempty"`
+}
+
+// UserDefinedNetworkSpec defines the desired state of UserDefinedNetworkSpec.
+// +union
+type UserDefinedNetworkSpec struct {
+	// topology describes network configuration.
+	// Allowed values are "Layer3", "Layer2", "LocalNet".
+	// Layer3 topology creates one logical switch per node - each with a different subnet - and a router interconnecting all the logical switches.
+	// Layer2 topology creates one logical switch shared by all nodes.
+	// LocalNet topology creates a cluster-wide logical switch connected to a physical network.
+	// +kubebuilder:validation:Required
+	// +unionDiscriminator
+	// +required
+	Topology NetworkTopology `json:"topology"`
+
+	// l3Config is the Layer3 configuration.
+	// +optional
+	Layer3 *L3Config `json:"layer3,omitempty"`
+
+	// l2Config is the Layer2 configuration.
+	// +optional
+	Layer2 *L2Config `json:"layer2,omitempty"`
+
+	// localNetConfig is the LocalNet configuration.
+	// +optional
+	LocalNet *LocalNetConfig `json:"localNet,omitempty"`
```

Comment:
The `Topology` field is now redundant since each topology has its own field, right?

In general I like this model better, but it seems spec could end up being empty in case non of the typologies field  are specified (Layer2/Layer3/Topology), right?

The spec should at least specify the topology and role, maybe the Role field should be exported out from the topology config types.

How about the following structure:
```
type UserDefinedNetworkSpec struct {
  // +kubebuilder:validation:Required
  // +require
  Role NetworkRole

  // +kubebuilder:validation:Required
  // +require
  // +union
  Topology NetworkTopology
} 

+unionDiscriminator
type NetworkTopology struct {
        // l3Config is the Layer3 configuration.
	// +optional
	Layer3 *L3Config `json:"layer3,omitempty"`

	// l2Config is the Layer2 configuration.
	// +optional
	Layer2 *L2Config `json:"layer2,omitempty"`

	// localNetConfig is the LocalNet configuration.
	// +optional
	LocalNet *LocalNetConfig `json:"localNet,omitempty"`
}
```
Role and Topology are the only field in the spec and marked as required.

The NetworkTopology type serves as a union, all fields are optional, only one should allowed.
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#unions
I think it can be enforced with CEL rule using exist_one macro [1], for example:
// +kubebuilder:validation:XValidation:rule="self.topology.exist_one(layer2, layer3, localnet)", message="..."

WDYT?

[1] https://github.com/google/cel-spec/blob/v0.7.0/doc/langdef.md#macros

Reply:
I like your idea Or. It makes sense to me.

Reply:
I have used the format recommended by @JoelSpeed from https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions
I am sure he can better explain why having a discriminator is a good idea.

Reply:
There are 2 things being discussed here: 
- should the `role` attribute be part of the `UserDefinedNetworkSpec` (or within all the topology configurations)
- should we have a discriminator in `UserDefinedNetworkSpec`

I would have the role in `UserDefinedNetworkSpec`, as Or suggests. In localnet we can always hard-code to secondary for instance.

Regarding the discriminator, I think we should follow that best practice. I also like the fact it separates between topology (the string we use plenty in the OVN-K code) and topology configuration.

Reply:
about `role`: I am not sure why it should be outside of the topology spec, considering allowed `role` values depend on the topology type, which is exactly the point of having discriminated sets in the first place.

Reply:
> Specifying the topology config filed (l2/l3/localnet) implies the chosen topology naturally making the topology field redundant.

Your suggestion turns what is intended to be a declarative API into an implicit one. How would you work out the user intent if they had specified:
```
spec:
  layer2:
    ...
  layer3:
    ...
```
Having the required topology field makes it clear what the user intended, they have to make a choice, and then if they specify an incorrect struct underneath, we can validate that easily. 

This pattern is common across Kube now

Reply:
Hmm but that's not what Or suggested. Your example would need to be:

```
spec:
  topology:
     layer2:
```

what's wrong with this?

Reply:
Sorry, then my counter example is that I could create the following, and it's ambiguous as to my intent. With the discriminant, it's clear because we can tell the user "layer3 is forbidden when discriminant is not Layer3" for example
```
spec:
  topology:
    layer2:
      ...
    layer3:
      ...
```

Reply:
you cant provide multiple values to topology, it only takes a single value.

Reply:
The intention is to have only one option set at a time, having the spec describe the desired topology config explicitly by picking the option.
I followed the example described here [[1]](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#unions), it mention pod volume API [[2]](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#volume-v1-core) as an example, it has more then 10 option but only one can be set at a time, I do understand its pretty old API and I suppose some lessons have learned.
Now I realize OCP API conventions [[3]](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions) says the union discrimination shall have a field that describes the desisted option alongside the option details.

I still think `Role` field should be extract and be part of the top-level spec.
IMO it make more sense to describe the user-defined network spec as whole and not as a topology config detail.
If we go with extracting `Role` field, I think the union should be the `NetworkTopology` field instead of `spec`: 

<details>

  <summary><b> For example: (click me) </b></summary>

	type UserDefinedNetworkSpec struct {
		// +required
		Role NetworkRole `json:"role"`

		// +union
		// +required
		Topology NetworkTopology `json:"topology"`
	}

	// +kubebuilder:validation:Enum=Primary;Secondary
	type NetworkRole string

	const (
		NetworkRolePrimary   NetworkRole = "Primary"
		NetworkRoleSecondary NetworkRole = "Secondary"
	)

	// +unionDiscriminator
	type NetworkTopology struct {
		Type NetworkTopologyType

		// +optional
		Layer3 *Layer3Config `json:"layer3,omitempty"`

		// +optional
		Layer2 *Layer2Config `json:"layer2,omitempty"`

		// +optional
		LocalNet *LocalNetConfig `json:"localNet,omitempty"`
	}

	// +kubebuilder:validation:Enum=Layer2;Layer3;LocalNet
	type NetworkTopologyType string

	const (
		NetworkTopologyLayer3   NetworkTopologyType = "Layer3"
		NetworkTopologyLayer2   NetworkTopologyType = "Layer2"
		NetworkTopologyLocalNet NetworkTopologyType = "LocalNet"
	)

</details>

[1] https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#unions
[2] https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#volume-v1-core
[3] https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions

Reply:
> you cant provide multiple values to topology, it only takes a single value.

How are you validating that, and what error will the user be presented with when they set multiple and create an invalid spec?

With the discriminated union, if they set the `topology: Layer2`, and then set `layer3: ...`, they would receive an error "layer3 is forbidden when topology is not layer3", giving them a clear indication they should change something so that we understand their intent. They said they wanted layer2, but are now confusing it with layer 3. The same error message would appear if they had the layer2 struct or not, when the layer3 struct is populated.

Reply:
> I still think Role field should be extract and be part of the top-level spec.

If you extract it, you have to have validation that ensures it's not set to a particular value when the toplogy is local net right? Having it inside the union takes that away and makes it clear from the outset it cannot be set with localnet

Reply:
> How are you validating that, and what error will the user be presented with when they set multiple and create an invalid spec?

I think it can be validated using CEL rule, although the rule end up pretty combersome, for[ example](https://playcel.undistro.io/?content=H4sIAAAAAAAAA71Ty2rDMBD8lakKIYbimOZUX3PqP%2FiyyBtZVF4ZS0kJTf%2B9cpMUCrVIH1S3YWdndpbVi9LsVK1WK2w61k8BdovYMbSXSFZ4hO3JcACNjEjGcIutd84%2FWzHvzMA9SbQaex6D9VI20siyo7AMA%2Bsy%2BsE7bw6lowOP9wUWC9zMVNezVa%2FJCceiwPGIefl5gbz5FfIXys8MTsMV6k61FOlRhl1MW7%2BFlT052ybPqaFuBOld2s4QOIl%2FQKCPuxoPVVV9YqyzjHOArzmN%2FM8sf%2BqTTTTBye23Xt%2FSyKS%2BSiOTKJ1O71tOVzP92Nc3TeLN5bkDAAA%3D).
The CEL rule message should say spec is invalid because more then one of the union-elements was spesifed.
The error message will mention the discriminator field being misconfigured but not the wrongly specified fields (layer2 & layer3) though.

Nevertheless, we should use the best practice which is OCP convention for discriminated unions.

> If you extract it, you have to have validation that ensures it's not set to a particular value when the toplogy is local net right?  Having it inside the union takes that away and makes it clear from the outset it cannot be set with localnet

I see the advantage where validation is right next to the localnet related type.
**IMO** the structure make more sense having `Role` placed right under Spec.
In case to Role is extracted, similar validation rule with be defined the top of the Spec definition (instead of the LocalNetConfig type definition)

I you (and/or other reviewers) insist I dont mind having Role stay as is.

Reply:
using best practices is what we like to do, I hope we all agree that discriminator makes sense in the config now